### ISSUE TITLE
Experiment with using lossless JPEG-XL (colorspace=YUV 400)

### ISSUE THREAD

**Issue by JackKelly** [openclimatefix/Satip#67]
## Detailed Description
JPEG-XL is the "new kid on the block" for image compression. And it can losslessly compress greyscale images (using colorspace YUV 400). It might be great for compressing satellite images into our Zarr datasets.

## Context
See issue #45 for previous experiments and notes.

This great comparison of lossless compression using JPEG-XL, WebP, AVIF, and PNG suggests JPEG-XL wins. 

## Possible Implementation
`imagecodecs` includes an interface to JPEG-XL.

Next step is to try manually compressing images using the `cjxl` app (not ImageMagick). 

~~If that looks good then create a stand-alone little adapter library to adapt `imagecodecs` to be used with Zarr.  [Here's a super-simple little python library](https://github.com/glencoesoftware/zarr-jpeg2k/blob/master/zarr_jpeg2k/zarr_jpeg2k.py) (just 51 lines of code!) which enables jpeg-2000 compression in Zarr using imagecodecs.  Maybe it'd be possible to use the same pattern, but for JPEG-XL?~~ UPDATE: @cgohike has already [implemented this in imagecodecs](https://github.com/cgohlke/imagecodecs/blob/ddddb7bb0e34aa243e1b45c04b545343c0569828/imagecodecs/numcodecs.py#L527-L574)! (See comment below)

To use ImageMagick for quick experiments at the commandline:

You need [ImageMagick version > 7.0.10](https://jpegxl.io/tutorials/imagemagick/) to use JPEG-XL.

To install ImageMagick >7.0.10:

```shell
sudo apt-get install imagemagick libmagick++-dev 
```

See [here for how to install ImageMagick from source](https://linoxide.com/install-latest-imagemagick-on-ubuntu-20-04/).

Then use 'magick' command not 'convert'.

I'll do some experiments later today or tomorrow :slightly_smiling_face: 

## TODO

- [ ] investigate whether the JPEG-XL lossly uint8 images are fine as is.  It looks great visually. And 4 timesteps of UK HRV is only 0.6 MB (compared to 2.9 MB with bzip2; and 2.2 MB with JPEG-XL lossless uint16).
- [ ] investigate whether we can use different color profiles.  See [these docs](https://github.com/libjxl/libjxl/blob/v0.6.1/lib/include/jxl/encode.h#L166-L190).
- [x] investigate if imagecodecs JPEG-XL can simply be installed through pip (or does it require libjxl to be manually installed first?) If it requires manual install then that maybe makes it inappropriate for a dataset that we might release publicly?
- [x] to get to 8bits, divide by 4 AND ROUND
- [ ] Try the ideas [Jon suggested on the JPEG-XL github issue queue](https://github.com/libjxl/libjxl/issues/1137) (especially putting all 12 channels into a single JPEG-XL, using lossless compression)
- [x] See if it's possible to change the suffix of each Zarr chunk to `.jxl`  (can't see how to do this and, anyway, Chrome cannot currently open jxl files)
- [x] Check output is the same (or roughly the same) as the input: plot gamma curve; compute MSE; etc.
- [ ]  try using alpha channel for NaNs. (haven't tried this but not going to bother because we can just use float16)
- [x]  try float16 for saving NaNs.  Yup, float16 saves NaNs, and there's no gamma curve. Need to map values to the range [0, 1].
- [ ] try float32 with jpeg-xl
- [ ] measure decompression speed of jpeg-xl vs gzip2
- [ ] prepare a PR for Satip for using jpeg-xl

**Comment by cgohlke** [comment]
> create a stand-alone little adapter library to adapt imagecodecs to be used with Zarr

Try https://github.com/cgohlke/imagecodecs/blob/ddddb7bb0e34aa243e1b45c04b545343c0569828/imagecodecs/numcodecs.py#L527-L574

**Comment by JackKelly** [comment]
Awesome, thanks so much for your help!

**Comment by JackKelly** [comment]
Some initial result compressing a single satellite image (`HRV_MSG3-SEVI-MSG15-0100-NA-20190620105917_cropped.png`) using `cjxl --quality=100`:

| effort | runtime | filesize (MB) |
|--------|---------|---------------|
| 1      | 0.7 sec | 6.6           |
| 7      | 20 sec  | 4.9           |
| 9      |         | 4.8           |

**Comment by JackKelly** [comment]
reading through the jpeg-xl code in imagecodecs, it looks like all we have to do is cast our `np.int16` images to `np.uint16` and pass the images in, one satellite channel at a time, and the code will figure out that it should use grayscale :slightly_smiling_face:

**Comment by JackKelly** [comment]
https://github.com/cgohlke/imagecodecs/issues/29

**Comment by JackKelly** [comment]
OK, some initial findings about using JPEG-XL with Zarr:

For `jpegxl_encode` to recognise the first dimension as a time dim, it's necessary to append a length 1 channel dim as the last dim; and convert to uint16 or uint8:  e.g. `satellite_dataset.example_dims(dim="channe", axis=-1).clip(min=0).astype(np.uint16)`

To engage lossless mode, use `JpegXL(distance=0, level=None, effort=<int between 1 and 9>)`. At the moment, `numcodecs.jpegxl_encode` enables lossless mode if and only if `level` is None or < 0 (which I think might be a bug: `level` actually sets the **decoding** speed. I've submitted a bug [here](https://github.com/cgohlke/imagecodecs/issues/30)).

Using lossless mode is a little fiddly.  To engage lossless mode, `level` must not be None.  When using `uint16` values in the range [0, 1023], I think `JPEG-XL` assumes the image is really dark, and compresses the hell out of it, unless `distance` is tiny (like 0.01).  `distance` can be 0 or 0.01, it can't be non-zero and less than 0.01.  If using values in the range [0, 1023], and `distance=0.01`, then the output values are in the wrong range, although the image looks OK.

Expanding the range of values to the range [0, 2**16-1] means that `distance` can be more like `1`. But it looks like it changes the gamma of the image, which is probably bad for our use-case.  Here's the compressed image:

![image](https://user-images.githubusercontent.com/460756/151603291-c88d0a95-2fa5-460f-a29f-bd9babce716f.png)

And the raw uncompressed image:

![image](https://user-images.githubusercontent.com/460756/151603319-a5e53294-15aa-4aa6-b8f6-0c2743cb4030.png)

**Comment by JackKelly** [comment]
imagecodecs JPEG-XL can simply be installed through pip (nice!)  No need for libjxl to be installed manually as well.  That's great news because it means that enabling JPEG-XL compression in Python is as simple `pip install imagecodecs` :slightly_smiling_face:

**Comment by JackKelly** [comment]
Some conclusions:

* TL;DR: Let's use jpeg-xl, distance=0.6, effort=8, float16. Need to map values to the range [0, 1], and use `JpegXlFuture` wrapper.
* In the numbers above, JPEG-2000 actually might look a tiny bit better than JPEG-XL! But the imagecodecs implementation of JPEG-2000 doesn't appear to support float16 or float32 (so no representation of NaNs).

Detail results:

For four timesteps:

| algo | runtime | size | MAE | PSNR |
|------|---------|-----------|---|----|
| bz2, level=5, int16 | 6.8 sec | 2.9 MB | | |
| jpex-xl, distance=0, effort=7, uint16 | 8.15 sec | 2.2 MB | | |
| jpex-xl, distance=0, effort=9, uint16 | 29 sec | 2.2 MB | 0 | |
| jpex-xl, distance=0, effort=9, float16 | 27.5 sec | 2.2 MB | 0 | 100 |
| jpex-xl, distance=1, effort=7, uint16 | 8.2 sec | 2.2 MB | | |
| jpeg-xl, distance=0.5, effort=9, uint8 | 16.2 s | 0.6 MB | 0.378 | 51.27 |
| jpeg-xl, distance=0.5, effort=8, uint8 | 7.8 s | 0.6 MB | 0.379 | 51.26 |
| jpeg-xl, distance=0.5, effort=7, uint8 | 6.8 s | 0.6 MB | 0.365 | 51.51 |
| jpeg-xl, distance=0.0, effort=7, uint8 | 8.58 s | 1.3 MB | | |
| jpeg-xl, distance=0.0, effort=9, uint8 | 18.2 s | 1.3 MB | | |
| jpeg-xl, distance=1.0, effort=9, uint8 | 12.8 s | 0.4 MB | 0.593 | |
| jpeg-xl, distance=0.1, effort=9, uint8 | 35.6 s | 1.8 MB | | |
| jpeg-xl, distance=0.2, effort=8, uint8 | 10.6 s | 1.2 MB | 0.136 |56.73 |
| jpeg-xl, distance=0.3, effort=8, uint8 | 7.8 s | 0.86 MB | 0.242 | 53.99 |
| jpeg-xl, distance=0.4, effort=8, uint8 | 8.08 s | 0.69 MB | 0.317 | 52.45 |
| jpeg-xl, distance=0.6, effort=8, float16 | 7.8 s | 0.74 MB  | 0.376; or 0.353 after rounding | 52.34 |
| jpeg-xl, distance=0.5, effort=7, float16 | 5.43 s | 0.87 MB | 0.324 | 53.85 |
| jpeg-xl, distance=0.4, effort=8, float16 | 7.8 s | 1.1 MB | 0.2749 | 55.17 |
| jpeg-2k, level=100, uint8 | 4.9 s | 1.1 MB | 0.1625  | 56.01 |
| jpeg-2k, level=80, uint8 | 5.2 s | 1.1 MB | 0.163 | 55.99 |
| jpeg-2k, level=65, uint8 | 5.0 s | 1.0 MB | 0.178 | 55.55 |
| jpeg-2k, level=60, uint8 | 5.0 s | 0.9 MB | 0.210 | 54.78 |
| jpeg-2k, level=55, uint8 | 5.0 s | 0.75 MB | 0.2935 | 53.01 |
| jpeg-2k, level=50, uint8 | 4.9 s | 0.5 MB | 0.468 | 49.79 |
| bz2, level=5, uint8 | 5.64 sec | 1.8 MB | | |

Some plots of jpeg-xl, distance=0.6, effort=8, float16:

The compressed image:
![image](https://user-images.githubusercontent.com/460756/152023048-360f7b52-f80c-43cb-a096-61dd8b36902c.png)

![image](https://user-images.githubusercontent.com/460756/152023216-03913e30-219b-4848-af81-c439e6b1797c.png)

![image](https://user-images.githubusercontent.com/460756/152023473-77cde7d9-92fd-4ae9-992f-ef7540e81d5a.png)

![image](https://user-images.githubusercontent.com/460756/152023536-dd3cb60d-3f29-4619-88f5-549cf03878d3.png)

![image](https://user-images.githubusercontent.com/460756/152023631-8911322f-3add-405f-9319-f0ac3ddac863.png)

**Comment by JackKelly** [comment]
Some notes from today's experiments...

## JPEG-XL compression for our pre-prepared batches

Our "pre-prepared batches" have these dimensions:

1. example: 32
2. time: 24
3. y: 24
4. x: 24
5. channels: 11

AFAICT, NetCDF cannot use the numcodecs API. And so we cannot use JPEG-XL (or any other imagecodecs) to compress data into NetCDF files. (Which is a shame because it might have been nice to use JPEG-XL to compress our pre-prepared batches.)

We could use Zarr for our pre-prepared batches. But the `JpegXl` class in `imagecodecs` only really works with images which are of shape `(y, x)`. We can make this work like this:

```python
class JpegXlFuture(JpegXl):
    """Simple hack to make the JpegXl compressor in the currently released
    version of imagecodecs (version 2011.11.20) look like the version in development
    
    See https://github.com/cgohlke/imagecodecs/issues/31#issuecomment-1026179413
    """
    
    codec_id = 'imagecodecs_jpegxl'
    
    def __init__(self, lossless=None, decodingspeed=None, level=None, distance=None, *args, **kwargs):
        """
        Args:
            distance: Lowest settings are 0.00 or 0.01.  If 0.0 then also set lossless to True.
            level: DON'T SET THIS WITH THIS JpegXlFuture wrapper! 
                In imagecodecs version 2011.11.20, level is mapped (incorrectly) to the decoding speed tier. 
                Minimum is 0 (highest quality), and maximum is 4 (lowest quality). Default is 0.
        """
        assert decodingspeed is None
        print(f"{lossless=}")
        if lossless:
            assert distance is None or distance == 0
            assert level is None  # level must be None to enable lossless in imagecodecs 2011.11.20.
        elif lossless is not None:
            level = 0  # level must be set to 0, 1, 2, 3, or 4 to enably lossy compression in imagecodecs 2011.11.20.
        super().__init__(level=level, distance=distance, *args, **kwargs)
        
    def encode(self, buf):
        n_examples = buf.shape[0]
        n_timesteps = buf.shape[1]
        n_channels = buf.shape[-1]
        assert n_examples == 1
        assert n_timesteps == 1
        assert n_channels == 1
        buf = buf[0]
        return super().encode(buf)
    
    def decode(self, buf, out=None):
        assert out is None
        out = super().decode(buf, out)
        out = out[np.newaxis, ...]
        return out
        
register_codec(JpegXlFuture)

# Convert to uint8
satellite_dataset["data"] = (satellite_dataset["data"].clip(min=0, max=1023) / 4).round().astype(np.uint8)

lossless = False
distance = 0.6
effort = 8

encoding = {
    "data": {
        "compressor": JpegXlFuture(lossless=lossless, distance=distance, effort=effort),
    },
}

chunks = dict(example=1, time_index=1, y_geostationary_index=None, x_geostationary_index=None, channels_index=1)

satellite_dataset.chunk(chunks).to_zarr(
    DST_PATH / (filename.stem + ".zarr"),
    mode="w",
    encoding=encoding,
    compute=True,
)
```

But this takes a whole minute per batch, and you end up with lots of _tiny_ Zarr chunks on disk (a few tens of _bytes_!)

The JPEG-XL spec allows for multiple images within a single JPEG-XL file. But I don't think `imagecodecs.JpegXl` knows how to decode multiple images. (Although I think it does know how to _encode_ multiple images!) When _encoding_ all timesteps into a single image, it "only" takes 20 seconds per batch.

**Comment by JackKelly** [comment]
## Using JPEG-XL for intermediate Zarrs

I'm much more optimistic about using JPEG-XL for our intermediate Zarrs.

The slightly fiddly thing is how to encode "NaNs".

JPEG-XL supports `uint8`, `uint16`, `float16`, and `float32`.

Trying to compress a `float16` or `float32` image which includes any NaNs results in JPEG-XL silently failing to save any image data to disk (not surprisingly: NaNs aren't really a thing in "proper" photography, which is what JPEG-XL is mostly designed for).

When using `float16` or `float32`, the image values are supposed to be in the range `[0, 1]`. It appears it _is_ possible to compress images where some values are outside that range (e.g. `-1`). But the "illegal" values change quite dramatically. And I'm keen to not do anything that's out-of-spec, for fear that it breaks something.

Where I've landed is for "real" image values to be in the range [0.075, 1], and use `0.025` to encode NaNs. Like this:

![image](https://user-images.githubusercontent.com/460756/153640789-2ebddd58-396a-4fb4-8a87-9a3b2914119f.png)

This does result in a bit of "ringing" (for want of a better term) around the edges of the "NaN box":

![image](https://user-images.githubusercontent.com/460756/153640937-a3af30ab-e55e-462a-a7d7-3f92c363ee9a.png)

But all values in the "NaN box" are between `0.0227` to `0.0280`.

(The ringing is _much_ larger amplitude if we use 0.95 as the "NaN box". Then the "NaN box" extends from about `0.90` to `0.99`)

Here's the code I'm using:

```python
original_float32 = original_float32.astype(np.float64)  # Use float64 for maths. Then turn to float32.

# Encode "NaN" as 0.025. Encode all other values in the range [0.075, 1].
# But, after JPEG-XL compression, there is slight "ringing" around the edges.
# So, after compression, use "image < 0.05" to find NaNs.
LOWER_BOUND_FOR_REAL_PIXELS = 0.075
NAN_VALUE = 0.025
original_float32 = original_float32.clip(min=0, max=1023) / (1023 * (1 + LOWER_BOUND_FOR_REAL_PIXELS))
original_float32 += LOWER_BOUND_FOR_REAL_PIXELS
original_float32 = original_float32.where(
    cond=original >= 0, 
    other=NAN_VALUE,
)
original_float32 = original_float32.astype(np.float32)

lossless = False
distance = 0.4
effort = 8

encoding = {
    "data": {
        "compressor": JpegXlFuture(lossless=lossless, distance=distance, effort=effort),
    },
}

# Need to set to a channel so that JpegXl interprets the first dimension as "time".
# original_float32 has chunks: `time=1, y=None, x=None`
zarr_store = original_float32.expand_dims(dim="channel", axis=-1).to_zarr(
    OUTPUT_ZARR_PATH,
    mode="w",
    encoding=encoding,
    compute=True,
)
```
